helpers/useRedirectUnauthenticated.ts

```diff
@@ -0,0 +1,23 @@
+import { useEffect } from 'react'
+import { useRouter } from 'next/router'
+import { LOGIN_PATH } from '../constants'
+
+/**
+ * Redirects to login page if user is not authenticated.
+ * With next query param to redirect back to this page after login.
+ * @param {boolean} shouldRedirect Condition for redirection
+ */
+const useRedirectUnauthenticated = (shouldRedirect: boolean) => {
+  const router = useRouter()
+
+  useEffect(() => {
```

Comment:
I think this should work without useEffect. Can you try removing useEffect and try to have it work without it? Because in the files that you refactored, useEffect is not used to handle this redirection logic for users who are not logged in. So it seems like we may be adding unnecessary logic by using useEffect here

Reply:
It'll work without useEffect, I used useEffect in my suggestion to not do it on every render which is what happens now.

Reply:
For this use case, useEffect is not a good choice.

Using useEffect for this use case adds an additional layer of complexity and makes it harder to maintain because components will be forced to do a re-render in order for the redirection logic to get run. There should be no dependency on needing to re-render components when redirecting users, it doesn’t make sense. Its an additional layer of complexity that is unnecessary.

Reply:
I'm not sure I understand you.

This useEffect is not changing state, it will not cause any extra re-renders. This code runs at the same time as not using useEffect when the redirection has to occur, and does not the rest of the time. The body of this useEffect will at most, in the worst case, run as many times as not using useEffect, and in the typical case fewer times, depending on the condition used to decide wether to redirect.

This redirection is happening client side, after the app state changes. On this change a re render will happen with or without this useEffect and the redirection is happening after this render starts, currently (without) as well with the useEffect that is not avoidable. This render does not complete though in either case since as long the page location is changed the execution of the current page will halt.

This code is performing a side effect when a state changes, exactly what useEffect is meant for.

Some examples:
https://nextjs.org/docs/api-reference/next/router#usage
https://github.com/nextauthjs/next-auth/discussions/4078#discussioncomment-2270608

This useEffect can be removed still if it is too complicated and redirection will work, but useEffect is not changing when the redirection is happening/order of execution of its body. Redirection is happening after a re-render without it too since it can only happen after the state changes as that is prerequisite to deciding wether to redirect on the client.

Reply:
useEffect only gets called on re renders. So if you implement this redirection logic with useEffect, its going to be dependent on needing a re-render in order for the redirection logic to kick in. It does not make sense to have our redirection logic be dependent on re-renders. Its an additional layer of complexity that provides no benefit

Reply:
I see. Yeah sure this will happen after the render is done. But the render will just be the loading spinner the same as before. The benefit is not running on every render, but if you feel that is not worth the complexity then sure.

Reply:
`The benefit is not running on every render`

We actually are not saving any extra pieces of logic from being run. After rendering finishes, useEffect will check the list of dependency values against the values from the last render, and will call your callback function if any one of them has changed. So the logic of checking still happens. Except it happens after every render rather than during.

Instead of `if (unauthenticated) router.push(...)` on every render

it'll be 
```
if (unauthenticated dependency has changed) {
 Do callback for redirection logic
}
```

after every render.